### PR TITLE
GH-167: Add to home screen banner in order confirmation

### LINKED ISSUE #167: Implement Add to Home in checkout confirmation page
See #160 for wireframes
PWA add to home component implemented already
Let's add it to checkout confirmation page

### PR DESCRIPTION
**Pull request by marlass**
Create service for add to home screen feature. New banner with prompt in the order confirmation.
Closes #167

### CODE REVIEW

File: projects/storefrontlib/src/lib/pwa/add-to-home-screen-btn/add-to-home-screen-btn.component.spec.ts [hunk 1]
```diff
@@ -2,15 +2,19 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/pwa/add-to-home-screen-btn/add-to-home-screen-btn.component.spec.ts]: I made some changes on this test, and pushed my changes.

File: projects/storefrontlib/src/lib/pwa/add-to-home-screen-btn/add-to-home-screen-btn.component.html [hunk 2]
```diff
@@ -1,3 +1,3 @@
 <span (click)="prompt()">
-  <ng-content *ngIf="isEnabled && canPrompt"></ng-content>
+  <ng-content *ngIf="canPrompt$ | async"></ng-content>
 </span>
```
[the remark this answers is not in the document]
Comment (marlass) [reply projects/storefrontlib/src/lib/pwa/add-to-home-screen-btn/add-to-home-screen-btn.component.html]: No. We wanted to prompt on click for any children. That's why we wrapped `ng-content`

File: projects/storefrontlib/src/lib/pwa/pwa.module.ts [hunk 3]
```diff
@@ -28,7 +34,14 @@ export function pwaConfigurationFactory(
       provide: RegistrationOptions,
       useFactory: pwaConfigurationFactory,
       deps: [Config]
-    }
+    },
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/pwa/pwa.module.ts]: why not just call "init" in the service's constructor. So, this part can be removed.

File: projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import { Injectable } from '@angular/core';
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts]:
```
import { Injectable } from '@angular/core';
import { PWAModuleConfig } from '../pwa.module-config';
import { GlobalMessageType } from './../../global-message/models/message.model';
import { GlobalMessageService } from '../../global-message/facade/global-message.service';

@Injectable()
export class AddToHomeScreenService {
  private deferredEvent: any;
  canPrompt = false;

  constructor(
    private config: PWAModuleConfig,
    private globalMessageService: GlobalMessageService
  ) {
    if (this.config.pwa.addToHomeScreen) {
      this.init();
    }
  }

  init() {
    window.addEventListener('beforeinstallprompt', event => {
      event.preventDefault();
      this.deferredEvent = event;
      this.canPrompt = true;
    });

    window.addEventListener('appinstalled', () => {
      this.globalMessageService.add({
        type: GlobalMessageType.MSG_TYPE_CONFIRMATION,
        text: 'SAP Storefront was added to your home screen'
      });

      this.canPrompt = false;
      this.deferredEvent = null;
    });
  }

  firePrompt() {
    if (this.deferredEvent) {
      this.deferredEvent.prompt();
    }
  }
}
```
I rewrite this service. I feel it is more readable.

Comment (marlass) [reply projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts]: Looks better. Changed to this version, but kept canPrompt as observable.

File: projects/storefrontlib/src/lib/pwa/pwa.module.ts (hunk 3 above)
Comment (marlass) [reply projects/storefrontlib/src/lib/pwa/pwa.module.ts]: Removed `init` call after service refactor.

**Comment by marlass** [comment]
@WeizhengSap 
1. `AddToHomeScreenBannerComponent` is moved to `PwaModule`.
2. Extracted common part from `AddToHomeScreenBtnComponent` and `AddToHomeScreenBannerComponent` to separate file. These components now extends it. Also moved tests from `AddToHomeScreenBtnComponent` to this new one. Removed the tests for btn component because currently component don't have logic (only extends the generic one).
3. Applied refactor suggestion for service.

Could you check it again?

File: projects/storefrontlib/src/lib/pwa/components/add-to-home-screen.component.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { OnInit } from '@angular/core';
+import { AddToHomeScreenService } from './../services/add-to-home-screen.service';
+import { Observable } from 'rxjs';
+
+export class AddToHomeScreenComponent implements OnInit {
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/pwa/components/add-to-home-screen.component.ts]: can we make this class as 'abstract'?

File: projects/storefrontlib/src/lib/pwa/pwa.module.ts [hunk 6]
```diff
@@ -8,14 +8,21 @@ import {
...
+import { AddToHomeScreenBannerComponent } from './components/add-to-home-screen-banner/add-to-home-screen-banner.component';
+import { AddToHomeScreenService } from './services/add-to-home-screen.service';
 
 export function pwaConfigurationFactory(
   pwaConfig: PWAModuleConfig
 ): RegistrationOptions {
   return { enabled: (pwaConfig.production && pwaConfig.pwa.enabled) || false };
 }
 
+export function pwaFactory(addToHomeScreenService) {
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/pwa/pwa.module.ts]: Can we remove this pwaFactory and this 
```
{
      provide: APP_INITIALIZER,
      useFactory: pwaFactory,
      deps: [AddToHomeScreenService],
      multi: true
    }
```
Just provide AddToHomeScreenService should be ok, right?

File: projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import { GlobalMessageService } from '../../global-message/facade/global-message.service';
+import { BehaviorSubject } from 'rxjs';
+
+@Injectable()
+export class AddToHomeScreenService {
+  private deferredEvent: any;
+
+  private canPrompt = new BehaviorSubject<boolean>(false);
+
+  canPrompt$ = this.canPrompt.asObservable();
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts]: why we want an observable canPrompt$? A boolean variable is enough, right?

File: projects/storefrontlib/src/lib/pwa/pwa.module.ts (hunk 6 above)
Comment (marlass) [reply projects/storefrontlib/src/lib/pwa/pwa.module.ts]: It won't work. This service needs to be initialized on app initialization and should be created even when don't have any consumers. `beforeinstallprompt` event can be called in 1 second or after 1 minute. As far as I know this is the only solution.

File: projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts (hunk 7 above)
Comment (marlass) [reply projects/storefrontlib/src/lib/pwa/services/add-to-home-screen.service.ts]: Boolean isn't enough. Each consumer should react instantly when `beforeinstallprompt` event is fired. Button in a mobile menu is initialized on app start. The event can be triggered 30 second later and this component should be shown there right after.

File: projects/storefrontlib/src/lib/pwa/components/add-to-home-screen.component.ts (hunk 5 above)
Comment (marlass) [reply projects/storefrontlib/src/lib/pwa/components/add-to-home-screen.component.ts]: Good idea. Changed it.

**Pull request merged**